### ISSUE TITLE
Enhancement: throw an error for invalid `TSParameterProperty`s

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#11708]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [x] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

ast-spec

### My proposal is suitable for this project

- [x] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

```ts
class Test1 {
  constructor(
    private [foo] = [1],
    private { bar } = { bar: 1 },
    private ...baz
  ) { }
}
```

Currently these three parameter properties are semantic errors in TS and are similarly allowed by our parser.
However they are all non-sensical. The first two [don't even transpile to generate a property](https://www.typescriptlang.org/play/?#code/MYGwhgzhAEAqCmEAuBGaBvAUNawD2AdsgE4CuwSexAFNjtAA7ECWAbmEvNANoBmeeALrQAvDxSCANHRxM2HLumgAjMMWgBfURhVqAXNDQbp9Ri3adoAOhuqAXnQCUOjZldA)!

We should make these hard errors to block the syntax so that we can refine our [`TSParameterProperty.parameter`](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/ast-spec/src/parameter/TSParameterProperty/spec.ts) type to exclude the invalid states.

An additional case which is already a hard TS parser error that we should pre-narrow the types for -- the `.left` of an assignment parameter property _must_ always be an `Identifier`.

**Comment by bradzacher** [comment]
cc @fisker as I'm sure you'd like to be able to skip this case in prettier.

**Comment by fisker** [comment]
Yes, Prettier is fine with this change.

**Comment by ronakmaheshwari** [comment]
@bradzacher Can I work on this issue?

**Comment by bradzacher** [comment]
Yup
https://typescript-eslint.io/contributing/issues#issue-claiming

**Comment by ronakmaheshwari** [comment]
@bradzacher  completed the fix Can you please review it #11753
